libs/vkd3d/command.c

```diff
@@ -2955,6 +2956,114 @@ static HRESULT STDMETHODCALLTYPE d3d12_command_list_Close(d3d12_command_list_ifa
+    size_t hi = list->query_ranges_count;
+    size_t lo = 0;
+
+    while (lo < hi)
+    {
+        size_t pos = lo + (hi - lo) / 2;
+        range = &list->query_ranges[pos];
+
+        if (vk_pool < range->vk_pool)
+            hi = pos;
+        else if (vk_pool > range->vk_pool)
+            lo = pos + 1;
+        else if (index < range->index)
+            hi = pos;
+        else if (index >= range->index + range->count)
+            lo = pos + 1;
+        else
+        {
+            if (out_pos)
+                *out_pos = pos;
+            return true;
+        }
+    }
+
+    if (out_pos)
+        *out_pos = lo;
+    return false;
+}
+
+static void d3d12_command_list_insert_query_range(struct d3d12_command_list *list, size_t pos,
+        VkQueryPool vk_pool, uint32_t index, uint32_t count, uint32_t flags)
+{
+    struct vkd3d_query_range *range;
+
+    vkd3d_array_reserve((void**)&list->query_ranges, &list->query_ranges_size,
+            list->query_ranges_count + 1, sizeof(*list->query_ranges));
+
+    for (size_t i = list->query_ranges_count; i > pos; i--)
+        list->query_ranges[i] = list->query_ranges[i - 1];
+
+    range = &list->query_ranges[pos];
+    range->vk_pool = vk_pool;
+    range->index = index;
+    range->count = count;
+    range->flags = flags;
+
+    list->query_ranges_count++;
+}
+
+static void d3d12_command_list_read_query_range(struct d3d12_command_list *list,
+        VkQueryPool vk_pool, uint32_t index, uint32_t count)
+{
+    const struct vkd3d_query_range *range;
+    size_t hi = index + count;
+    size_t lo = index;
+    size_t pos;
+
+    /* pos contains either the location of an existing range
+     * containing the first query of the new range, or the
+     * location where we need to insert it */
+    d3d12_command_list_find_query(list, vk_pool, index, &pos);
+
+    /* Do not attempt to merge adjacent ranges, but make sure
+     * that each query is only contained in one range */
+    while (lo < hi)
+    {
+        if (pos < list->query_ranges_count)
+        {
+            range = &list->query_ranges[pos];
+
+            if (lo >= range->index)
+                lo = max(lo, range->index + range->count);
+            else
+            {
+                size_t range_end = min(hi, range->index);
+                d3d12_command_list_insert_query_range(list, pos,
+                        vk_pool, lo, range_end - lo, 0);
+                lo = range_end;
+            }
+
+            pos += 1;
+        }
+        else
+        {
+            d3d12_command_list_insert_query_range(list, pos,
+                    vk_pool, lo, hi - lo, 0);
+            lo = hi;
+        }
+    }
+}
+
+static bool d3d12_command_list_reset_query(struct d3d12_command_list *list,
+        VkQueryPool vk_pool, uint32_t index)
+{
+    size_t pos;
+
+    if (d3d12_command_list_find_query(list, vk_pool, index, &pos))
+        return false;
+
+    d3d12_command_list_insert_query_range(list, pos,
```

Comment:
What happens here if we read query index 2, and then later reset range [0, 4]. We would need to split this into three ranges. [0, 1] reset, 2 read, [3, 4] reset.

Reply:
We can only ever reset one query at a time (since this happens in `{Begin|End}Query`, so we can never override read queries.

Reply:
Right. I was wondering if it made sense to merge query ranges, say that the app is doing query 0, 1, 2, 3, ... in order. Ideally, we should be able to just bump the range instead of ending up with lots of different ranges, which doesn't seem ideal when we have a binary search in the hot path.

Reply:
I deliberately left that out since it would make the insertion code completely insane, since now we'd have to handle the case where inserting a range does not insert a new item, or even removes an item because it merges three adjacent ranges into one, and adjust our `pos` cursor in some way. I'm having a very hard time reasoning about all the special cases.

Binary search should never have more than 10-12 iterations per lookup anyway, and that is if the game uses an extreme number of queries.

I guess the bigger problem is that insertion is `O(n^2)`

Reply:
Ok, if it makes the code insane guess we can leave it on the table until it becomes a real problem.